### PR TITLE
#17 fixes Invalidation of buttons reaction by mouse events

### PR DESCRIPTION
**Pull request by skydivin4ng3l**
buttons created after mouse event will not be triggered by the event creating them e.g.

### CODE REVIEW

**Comment by skydivin4ng3l** [comment]
questionable if _Buttons array is still needed?

File: Buttons.d [hunk 1]
```diff
@@ -326,6 +305,29 @@ func int Button_GetCaptionPtr(var int hndl) {
 	return v.textLines_next;
 };
 
+func void _Button_processMouseEvents( var int btn_hndl ) {
```
Comment (szapp) [inline Buttons.d]: - Change function return type to `int` (for proper stack handling with `foreachHndl`)
- Rename function to `_Buttons_Do_Sub` for consistent style
- Remove white space between parentheses for consistent/readable code

```D
func int _Buttons_Do_Sub(var int btn_hndl) {
```

File: Buttons.d [hunk 2]
```diff
@@ -326,6 +305,29 @@ func int Button_GetCaptionPtr(var int hndl) {
...
+	var _Button btn; btn = get( btn_hndl );
```
Comment (szapp) [inline Buttons.d]: Remove white space between parentheses for consistent/readable code

```D
	var _Button btn; btn = get(btn_hndl);
```

File: Buttons.d [hunk 3]
```diff
@@ -326,6 +305,29 @@ func int Button_GetCaptionPtr(var int hndl) {
...
+				MEM_CallByID(btn.on_enter);
+				btn.state = btn.state | BUTTON_ENTERED;
+			};
+		} else if (btn.state & BUTTON_ENTERED) {
+			MEM_PushIntParam(btn_hndl);
+			MEM_CallByID(btn.on_leave);
+			btn.state = btn.state & ~BUTTON_ENTERED;
+		}; 
+	};
+};
```
Comment (szapp) [inline Buttons.d:333]: Add return for `foreachHndl`

```D
    return rContinue;
};
```

File: Buttons.d [hunk 4]
```diff
@@ -341,35 +343,9 @@ func void Buttons_Do() {
...
-				MEM_CallByID(btn.on_leave);
-				btn.state = btn.state & ~BUTTON_ENTERED;
-			}; 
-		};
-				
-		i += 1;
-		MEM_StackPos.position = pos;
+
+	foreachHndl( _Button@, _Button_processMouseEvents );
+
```
Comment (szapp) [inline Buttons.d]: Remove empty line

File: Buttons.d [hunk 5]
```diff
@@ -341,35 +343,9 @@ func void Buttons_Do() {
...
-				MEM_PushIntParam(MEM_ReadStatArr(_Buttons, i));
-				MEM_CallByID(btn.on_leave);
-				btn.state = btn.state & ~BUTTON_ENTERED;
-			}; 
-		};
-				
-		i += 1;
-		MEM_StackPos.position = pos;
+
+	foreachHndl( _Button@, _Button_processMouseEvents );
```
Comment (szapp) [inline Buttons.d]: - Adjust function name (see [comment in line 308](https://github.com/Lehona/LeGo/pull/18/files#diff-4a3ff6edbf0e7ae29312af1298113f4dR308))
- Remove white space between parentheses for consistent/readable code

```D
	foreachHndl(_Button@, _Button_Do_Sub);
```

**Review by szapp — CHANGES_REQUESTED** [review]
Forgive the picky feedback concerning code style. Note: I have not tested your changes.

File: Buttons.d (hunk 3 above)
Comment (skydivin4ng3l) [reply Buttons.d:333]: what is rContinue?

File: Buttons.d (hunk 2 above)
Comment (skydivin4ng3l) [reply Buttons.d]: only accept consistency as reason but not readability, think everybody has their own preference. ;)

File: Buttons.d (hunk 3 above)
Comment (szapp) [reply Buttons.d:333]: It's essentially `continue` with a different name (to not confuse them with the Ikarus jump symbols I assume):

https://github.com/Lehona/LeGo/blob/d76cbccf211a2816076a80d4bc7fd2d72253d694/PermMem.d#L183-L188

**Review by szapp — APPROVED** [review]

**Comment by szapp** [comment]
Thanks for the changes!

@Lehona any more thoughts? I think this looks good.

**Pull request merged**